rslib/src/collection/backup.rs

```diff
@@ -0,0 +1,477 @@
+    }
+
+    fn remaining(&self, stage: BackupStage) -> bool {
+        match stage {
+            BackupStage::Daily => self.limits.daily > 0,
+            BackupStage::Weekly => self.limits.weekly > 0,
+            BackupStage::Monthly => self.limits.monthly > 0,
+        }
+    }
+
+    fn keep_or_delete(&mut self, stage: BackupStage, backup: Backup) {
+        let keep = match stage {
+            BackupStage::Daily => backup.day() < self.last_kept_day,
+            BackupStage::Weekly => backup.week() < self.last_kept_week,
+            BackupStage::Monthly => backup.month() < self.last_kept_month,
+        };
+        if keep {
+            self.keep(Some(stage), backup);
+        } else {
+            self.delete(backup);
+        }
+    }
+
+    /// Adjusts limits as per the stage of the kept backup, and last kept times.
+    fn keep(&mut self, stage: Option<BackupStage>, backup: Backup) {
+        self.last_kept_day = backup.day();
+        self.last_kept_week = backup.week();
+        self.last_kept_month = backup.month();
+        match stage {
+            None => (),
+            Some(BackupStage::Daily) => self.limits.daily -= 1,
+            Some(BackupStage::Weekly) => self.limits.weekly -= 1,
+            Some(BackupStage::Monthly) => self.limits.monthly -= 1,
+        }
+    }
+
+    fn delete(&mut self, backup: Backup) {
+        self.obsolete.push(backup);
+    }
+}
+
+impl Meta {
+    fn from_archive(archive: &mut ZipArchive<File>) -> Option<Self> {
+        archive
+            .by_name("meta")
+            .ok()
+            .and_then(|file| serde_json::from_reader(file).ok())
+    }
+}
+
+fn check_collection(col_path: &Path) -> Result<()> {
+    let col = CollectionBuilder::new(col_path).build()?;
+    col.storage
+        .db
+        .pragma_query_value(None, "integrity_check", |row| row.get::<_, String>(0))
+        .map_err(Into::into)
+        .and_then(|s| match s.as_str() {
+            "ok" => Ok(()),
+            _ => Err(AnkiError::invalid_input(format!("corrupt: {}", s))),
+        })
+}
+
+fn restore_media(
+    mut progress_fn: impl FnMut(usize) -> bool,
+    archive: &mut ZipArchive<File>,
+    media_folder: &str,
+) -> Result<()> {
+    let media_file_names = extract_media_file_names(archive).ok_or(AnkiError::NotFound)?;
+    let mut count = 0;
+
+    for (archive_file_name, file_name) in media_file_names {
+        count += 1;
+        if count % 10 == 0 && !progress_fn(count) {
+            return Err(AnkiError::Interrupted);
+        }
+
+        if let Ok(mut zip_file) = archive.by_name(&archive_file_name) {
+            let file_path = Path::new(&media_folder).join(normalize_to_nfc(&file_name).as_ref());
+            let files_are_equal = fs::metadata(&file_path)
+                .map(|metadata| metadata.len() == zip_file.size())
+                .unwrap_or_default();
+            if !files_are_equal {
+                io::copy(&mut zip_file, &mut File::open(file_path)?)?;
```

Comment:
Since you're sure that media will be compresed in the future, it makes sense to handle that now, and make this code as forward compatible as possible.
I'll take care of it.

Reply:
But the current approach would still work, wouldn't it? I don't think speed should be a concern. Let's assume a user tries to open a .colpkg from a future Anki release with this one (presumably 2.1.50).
Option 1: Process it, sluggishly at worst.
Option 2: Refuse to process it.
The latter inconveniences the user much more.

Reply:
Just worried about committing to something and then finding out later than the raw compression outputs in a different format, eg without wrapping in frames or without some other header. If we can confirm they will be compatible, then agreed that option one is a much better choice.

Reply:
I had a quick glance, and it doesn't look like reusing decoding/encoding contexts is currently supported (https://github.com/gyscos/zstd-rs/issues/41). What about putting media files into a nested archive and compressing that?
Either way, it shouldn't be a big change, so I'd just implement whatever you think is the most likely scenario for the future, and minimise chances for breaking forward compatibility.

Reply:
After thinking about it a bit, I take that back. The archive approach would likely be somewhat more involved, and without having the generating counterpart, it would be hard to test.

Reply:
We can do it, but we need to drop down to a lower level:

```rust
use std::time::Instant;

use anyhow::Result;
use zstd::stream::raw::Encoder as RawEncoder;
use zstd::stream::zio::Writer;
use zstd::zstd_safe;
use zstd::Encoder;

const INPUT_BYTES: &[u8] = include_bytes!("main.rs");
const CYCLES: usize = 10000;

fn simple_compress() -> Result<Vec<u8>> {
    let mut reader = std::io::Cursor::new(INPUT_BYTES);
    let writer = Vec::new();
    let mut encoder = Encoder::new(writer, 0)?;
    encoder.multithread(16)?;
    std::io::copy(&mut reader, &mut encoder)?;
    encoder.finish().map_err(Into::into)
}

fn setup_reusable_encoder() -> Result<RawEncoder<'static>> {
    let mut encoder = RawEncoder::with_dictionary(0, &[])?;
    encoder.set_parameter(zstd_safe::CParameter::NbWorkers(16))?;
    Ok(encoder)
}

fn reusing_compress(encoder: RawEncoder<'static>) -> Result<(Vec<u8>, RawEncoder<'static>)> {
    let mut writer = Writer::new(Vec::new(), encoder);
    let mut reader = std::io::Cursor::new(INPUT_BYTES);
    std::io::copy(&mut reader, &mut writer)?;
    writer.finish()?;
    Ok(writer.into_inner())
}

fn main() -> Result<()> {
    // ensure we get identical results
    let data1 = simple_compress()?;
    let reusable_encoder = setup_reusable_encoder()?;
    let (data2, reusable_encoder) = reusing_compress(reusable_encoder)?;
    assert_eq!(data1, data2);
    let (data3, mut reusable_encoder) = reusing_compress(reusable_encoder)?;
    assert_eq!(data1, data3);

    // compare performance
    let start = Instant::now();
    for _i in 0..CYCLES {
        let (data, encoder) = reusing_compress(reusable_encoder)?;
        assert_eq!(data, data1);
        reusable_encoder = encoder;
    }
    println!(
        "reuse took {}",
        Instant::now().duration_since(start).as_millis()
    );

    let start = Instant::now();
    for _i in 0..CYCLES {
        let data = simple_compress()?;
        assert_eq!(data, data1);
    }
    println!(
        "simple took {}",
        Instant::now().duration_since(start).as_millis()
    );

    Ok(())
}
```

On a 50MB collection file with 100 cycles, the simple approach is actually a little faster:

reuse took 3933
simple took 3842

But when we process lots of little files, reusing the context is a big win:

reuse took 144
simple took 2697

Reply:
(because the output is identical, we can take a simple approach for decompression for now, and deal with encoding/optimizing in the future)

Reply:
One other point of interest - when turning off multithreading, the two cases are much closer together, and the reuse case is about twice as fast:

reuse took 68
simple took 147

Switching to a 1MB test file, single threading is still better performing than multithreading. At 10MB, multithreading is performing better than single threading. So we probably don't want to enable it for the media case, as media files tend to be smaller than 10MB (and we'd likely achieve better performance by decompressing multiple files in parallel if we wanted to push that any further).

Reply:
That looks great! Both for this PR and future import work.
Seems like I should have looked a level deeper as well, so to speak.